File: js/emit_numerators.js
function
emit_numerators(u, p)
{
	var q, v;

	v = {type:LINE, a:[], level:u.level};

	p = cdr(p);
	q = car(p);

	if (isnum(q)) {
		if (isdouble(q))
			emit_double(v, q);
		else if (Math.abs(q.a) != 1)
			emit_roman_text(v, Math.abs(q.a).toFixed(0));
		p = cdr(p);
	}

	while (iscons(p)) {

		q = car(p);
		p = cdr(p);

		if (isdenominator(q))
			continue;

		if (v.a.length > 0)
			emit_medium_space(v);

		emit_factor(v, q);
	}

	if (v.a.length == 0)
		emit_roman_text(v, "1"); // there were no numerators

	if (v.a.length == 1) {
		u.num = v.a[0];
		return;
	}

	emit_update(v);

	u.num = v;
}

File: app/emit_numerators.c
#include "app.h"

void
emit_numerators(struct atom *p)
{
	int n, t;
	char *s;
	struct atom *q;

	t = tos;

	n = count_numerators(p);

	p = cdr(p);
	q = car(p);

	while (iscons(p)) {

		q = car(p);
		p = cdr(p);

		if (!isnumerator(q))
			continue;

		if (tos > t)
			emit_medium_space();

		if (isrational(q)) {
			s = mstr(q->u.q.a);
			emit_roman_string(s);
			continue;
		}

		if (car(q) == symbol(ADD) && n == 1)
			emit_expr(q); // parens not needed
		else
			emit_factor(q);
	}

	if (t == tos)
		emit_roman_char('1'); // no numerators

	emit_update_list(t);
}
